### PR TITLE
feat(Core/Script): Playing music from GameObjects and added holiday zones music script

### PR DESCRIPTION
**Pull request by peti446**
## CHANGES PROPOSED:
-  Add TC commits:
https://github.com/TrinityCore/TrinityCore/commit/002b0a7a65167d15969d98b41af7f91d222df664
https://github.com/TrinityCore/TrinityCore/commit/963278459a1d9df5aa3c2e1c8b5f5f112f6a370c
https://github.com/TrinityCore/TrinityCore/commit/e3560f3f518ee8a76921b4922292805c03871d4a
https://github.com/TrinityCore/TrinityCore/commit/19f8acdb616909758391d1c1459fe61196531f3b
- Modified selection of events for brewfest to be able to play all events at SHATTRATH

## ISSUES ADDRESSED:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/3215

## TESTS PERFORMED:
- Builded with no errors
- For Brewfest event: started brewfest event and went to multiple locations to check music (stayed in each for 5 minutes)
- Pirate Day event: started the event and went to the location at booty bay to check music  (stayed for 5 minutes)
- Darkmoon Fair Music: started all fair events and checked music in all their locations  (stayed in each for 5 minutes)
- MidSummer event: started event and checked some location with horde and alliance character  (stayed in each for 5 minutes)

## HOW TO TEST THE CHANGES:

Brewfest event
1. Go to any brewfest event in any city (outside main entrance)
-- Some locations: 
--- Orgrimar: .go xyz 1278.01 -4398.94 26.31 1
--- Stormwind: .go xyz -9129.9 340.76 93.76 0
2. .event start 24
3. Check music

Pirate Day event
1. Go to the event location: .go xyz -14425 530 25 0
3. .event start 50
4. Check music

Darkmoon Fair Music
1. .event start 3
2. .event start 4
3. .event start 5
4. .go xyz -1570.13 146.31 -7.792 1
5. Check music
6. .go xyz -9545.3 93.82 58.94 0
7. Check music
8. .go xyz -2260 5211.64 -9.98 530
9. Check music

MidSummer event
1. .event start 1
2. Go to summer location for horde and aliance
-- Some locations:
--- For horde: .go xyz 143.54 -4723.56 17.24 1
--- For alliance: .go xyz -9377.34 16.67 62.66 0
--- PS: there are many more locations
3. Check music

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: data/sql/updates/pending_db_world/rev_1597618822007580900.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1597618822007580900');
+
+UPDATE `gameobject_template` SET `ScriptName`='go_midsummer_music' WHERE `entry`=188174;
```
Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_world/rev_1597618822007580900.sql:3]: you don't need multiple SQL files, just 1 is enough. Also please add an empty line at the end of it ;)

File: src/server/game/Entities/Object/Object.cpp [hunk 2]
```diff
@@ -2894,6 +2894,17 @@ void WorldObject::PlayDirectSound(uint32 sound_id, Player* target /*= NULL*/)
...
 
+
+void WorldObject::PlayDirectMusic(uint32 music_id, Player* target /*= NULL*/)
+{
+    WorldPacket data(SMSG_PLAY_MUSIC, 4);
+    data << uint32(music_id);
+    if (target)
+        target->SendDirectMessage(&data);
+    else
+        SendMessageToSet(&data, true);
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Object/Object.cpp:2905]: better to add brackets

File: src/server/scripts/World/go_scripts.cpp [hunk 3]
```diff
@@ -410,6 +410,366 @@ class go_heat : public GameObjectScript
...
+                        else if (rnd == 1)
+                        {
+                            _currentMusicEvent = EVENT_BREWFESTGOBLIN02;
+                            musicTime = EVENT_BREWFESTGOBLIN02_TIME;
+                        }
+                        else
+                        {
+                            _currentMusicEvent = EVENT_BREWFESTGOBLIN03;
+                            musicTime = EVENT_BREWFESTGOBLIN03_TIME;
+                        }
```
Comment (FrancescoBorzi) [inline src/server/scripts/World/go_scripts.cpp:506]: what about:

```
switch (rnd)
{
  case 0:
     _currentMusicEvent = EVENT_BREWFESTGOBLIN01;
     musicTime = EVENT_BREWFESTGOBLIN01_TIME;
     break;
  case 1:
    .....
}
```

File: src/server/scripts/World/go_scripts.cpp [hunk 4]
```diff
@@ -410,6 +410,366 @@ class go_heat : public GameObjectScript
...
+                        else if (rnd == 1)
+                        {
+                            _currentMusicEvent = EVENT_BREWFESTDWARF02;
+                            musicTime = EVENT_BREWFESTDWARF02_TIME;
+                        }
+                        else
+                        {
+                            _currentMusicEvent = EVENT_BREWFESTDWARF03;
+                            musicTime = EVENT_BREWFESTDWARF03_TIME;
+                        }
```
Comment (FrancescoBorzi) [inline src/server/scripts/World/go_scripts.cpp:524]: same here

File: src/server/scripts/World/go_scripts.cpp [hunk 5]
```diff
@@ -410,6 +410,366 @@ class go_heat : public GameObjectScript
...
+                        else if (rnd == 4)
+                        {
+                            _currentMusicEvent = EVENT_BREWFESTDWARF02;
+                            musicTime = EVENT_BREWFESTDWARF02_TIME;
+                        }
+                        else
+                        {
+                            _currentMusicEvent = EVENT_BREWFESTDWARF03;
+                            musicTime = EVENT_BREWFESTDWARF03_TIME;
+                        }
```
Comment (FrancescoBorzi) [inline src/server/scripts/World/go_scripts.cpp:558]: same here, very long if-else chain :D

File: data/sql/updates/pending_db_world/rev_1597618822007580900.sql (hunk 1 above)
Comment (peti446) [reply data/sql/updates/pending_db_world/rev_1597618822007580900.sql]: Okey, they are merged now, I also added the missing boxes to sattrath. To confirm is there a set guid I should have used for this, right now I have used the last free ones to add them based on a clean install.

**Comment by LukasVolgger** [comment]
Awesome, thx!

**Comment by pklloveyou** [comment]
Tested

**Comment by pklloveyou** [comment]
but @FrancescoBorzi Please add
```sql
DELETE FROM `game_event_gameobject` WHERE `guid` IN (@LASTGUID+1,@LASTGUID+2);
INSERT INTO `game_event_gameobject` VALUES (24, @LASTGUID+1);
INSERT INTO `game_event_gameobject` VALUES (24, @LASTGUID+2);
```

**Comment by Kitzunu** [comment]
Also I will look up a better GUID to use, so we can keep it as small and compact as possible

**Comment by peti446** [comment]
@pklloveyou Thanks for pointing it out, will add it later.

**Comment by peti446** [comment]
@Kitzunu How does guid 4719 and 4720 sound like (4719 til 4720 are free) or 4595-4596 or 4751-4752

**Pull request merged**